File: CusipPriceProcessor/CusipLatestPrice.cs

namespace CusipPriceProcessor
{
	public struct CusipLatestPrice
	{
		public string Cusip { get; }

		public decimal Price { get; }

		public bool HasValue { get; }

		public static CusipLatestPrice Null = new CusipLatestPrice(null);

		public CusipLatestPrice(string cusip, decimal price)
		{
			Cusip = cusip;
			Price = price;
			HasValue = true;
		}

		public CusipLatestPrice(string cusip)
		{
			Cusip = cusip;
			Price = decimal.MinValue;
			HasValue = false;
		}
	}
}

File: CusipPriceCpp/CusipLatestPrice.cpp
#include "CusipLatestPrice.h"


CusipLatestPrice::CusipLatestPrice(string cusip, double price /*= std::numeric_limits<double>::lowest()*/) :
	Cusip(cusip), Price(price)
{
	if (Null == nullptr)
		Null = std::make_shared<CusipLatestPrice>(new CusipLatestPrice(nullptr));

	HasValue = (price > std::numeric_limits<double>::lowest());
}

CusipLatestPrice::~CusipLatestPrice() { }
